runtime/onert/core/src/loader/train/CheckpointLoader.cc

```diff
@@ -76,10 +145,95 @@ class CheckpointLoader final
       _file.close();
   }
 
+  void updateTensor(const std::unique_ptr<exec::Execution> &exec)
+  {
+    uint32_t vindex = 0;
+    exec->iterateTrainableTensors(
+      [&](const ir::OperandIndex &, const backend::train::ITrainableTensor *) { vindex++; });
+
+    if (_header.length != vindex)
+      throw std::runtime_error{
+        "Invalid number of tensors between TrainingInfo and checkpoint file"};
+
+    // Reset EOF bit
+    _file.clear();
+
+    vindex = 0;
+    exec->iterateTrainableTensors(
+      [&](const ir::OperandIndex &, const backend::train::ITrainableTensor *tensor) {
+        assert(tensor);
+        assert(_tensor_data[vindex].size == tensor->total_size());
+        _file.seekg(static_cast<std::streamoff>(_tensor_data[vindex].offset), std::ios::beg);
+        _file.read(reinterpret_cast<char *>(tensor->buffer()),
+                   static_cast<std::streamsize>(tensor->total_size()));
+        vindex++;
+      });
+  }
+
+  void updateOptimizer(const std::unique_ptr<ir::train::TrainingInfo> &train_info,
+                       const std::unique_ptr<onert::exec::Execution> &exec)
+  {
+    ir::train::OptimizerCode ckpt_opt_code = ir::train::OptimizerCode::SGD;
+    // TODO Support other optimizers
+    if (_optimizer_data.size() == 2)
+      ckpt_opt_code = ir::train::OptimizerCode::Adam;
+
+    if (ckpt_opt_code != train_info->optimizerInfo().optim_code)
+      throw std::runtime_error{
+        "Not compatible optimizer type between TrainingInfo and checkpoint file"};
+
+    switch (train_info->optimizerInfo().optim_code)
+    {
+      case ir::train::OptimizerCode::Adam:
+        updateAdamOptimizer(exec);
+        break;
+      default:
+        break;
+    }
+  }
+
+  void updateTrainingInfo(const std::unique_ptr<ir::train::TrainingInfo> &train_info)
+  {
+    // TODO Verify cur_step value
+    train_info->trainingStep() = _footer.cur_step;
+  }
+
+private:
+  void updateAdamOptimizer(const std::unique_ptr<onert::exec::Execution> &exec)
+  {
+    // Adam optimizer has two optimizer variables. (mean, variance)
+    [[maybe_unused]] constexpr auto ADAM_VARIABLE_COUNT = 2;
+
+    // Reset EOF bit
+    _file.clear();
+
+    auto vindex = 0;
```

Comment:
(optional)
How about using an explicit type like the above `vindex`?
```suggestion
    uint32_t vindex = 0;
```